### DISCUSSION TITLE
"Caught handled exception, but response already started." when using sub apps

### DISCUSSION THREAD

**Discussion by mxab** [fastapi/fastapi#9098] [Questions] [answered]
### First check

* [x] I added a very descriptive title to this issue.
* [x] I used the GitHub search to find a similar issue and didn't find it.
* [x] I searched the FastAPI documentation, with the integrated search.
* [x] I already searched in Google "How to X in FastAPI" and didn't find any information.
* [x] I already read and followed all the tutorial in the docs and didn't find an answer.
* [x] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
* [x] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
* [x] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).
* [x] After submitting this, I commit to one of:
    * Read open issues with questions until I find 2 issues where I can help someone and add a comment to help there.
    * I already hit the "watch" button in this repository to receive notifications and I commit to help at least 2 people that ask questions in the future.
    * Implement a Pull Request for a confirmed bug.

### Example

Here's a self-contained, [minimal, reproducible, example](https://stackoverflow.com/help/minimal-reproducible-example) with my use case:

https://github.com/mxab/fastapi-subapp-exceptions-show-case

When I throw an exception in a mounted sub app I get a strange error:

> RuntimeError: Caught handled exception, but response already started.

I guess the question is if the exception_handler feature should be used like this, but the error is very confusing :)
Is there a better way?

```Python
class UnicornException(Exception):
    def __init__(self, name: str):
        self.name = name

app = FastAPI()

@app.get("/app")
def read_main():
    raise UnicornException(name="main")

foo = FastAPI()

@foo.get("/foo")
async def read_foo():
    raise UnicornException(name="foo")

app.mount("/foo", foo)

@app.exception_handler(UnicornException)
async def unicorn_exception_handler(request: Request, exc: UnicornException):
    return JSONResponse(status_code=418, content={"message": f"Oops! {exc.name}"},)
```

```Python
client = TestClient(app)


def test_read_main():
    response = client.get("/app")
    assert response.status_code == 418

# fails
def test_read_foo():
    response = client.get("/foo/foo")
    assert response.status_code == 418

```
### Description

When I run the test
```
poetry install
poetry run pytest tests
```

The test `test_read_foo` which triggers the exception in the sub app fails
```
FAILED tests/test_app.py::test_read_foo - RuntimeError: Caught handled exception, but response already started.
```

### Environment

* OS: []: Windows WSL1 / ubuntu
* FastAPI Version [e.g. 0.60.1]:

To know the FastAPI version use:
 

* Python version:

3.6.10

**Comment by hard-coders** [comment]
I got the same problem.
I had to get rid of sub-app and it works fine.

**Comment by pybt** [comment]
Just use @foo.exception_handler(UnicornException)

**Comment by hai-eb7** [comment]
So there is no way to capture the exception of sub-apps from the main app? I have several sub-apps but the error handling logic is the same for them, I don't think register for each app is convenient.

**Comment by haomingyin** [comment, marked as answer]
@hai-eb7  I had the same problem. and thanks for the thread. It seems sub-app exceptions are not handled by the main app.

What I did is register handlers to all sub-apps and it worked.
```python

# Exception handlers
def add_exception_handlers(_app: FastAPI):
    @_app.exception_handler(ApiAuthException)
    async def api_auth_exception_handler(request: Request, exc: ApiAuthException):
        return await handler.api_auth_exception_handler(request, exc)

    @_app.exception_handler(ApiException)
    async def api_exception_handler(request: Request, exc: ApiException):
        return await handler.api_exception_handler(request, exc)

add_exception_handlers(main_app)
add_exception_handlers(sub_app)
```
